src/sparseml/pytorch/image_classification/utils/trainer.py

```diff
@@ -0,0 +1,324 @@
+    ModuleTester,
+    ModuleTrainer,
+    default_device,
+    model_to_device,
+)
+
+
+_LOGGER = logging.getLogger(__file__)
+
+__all__ = [
+    "ImageClassificationTrainer",
+]
+
+
+class Trainer(ABC):
+    """
+    Abstract class for Trainers.
+    Creates a contract that all trainers must have a run_one_epoch method.
+    """
+
+    @abstractmethod
+    def run_one_epoch(self):
+        """
+        Runs one epoch of training.
+        """
+        raise NotImplementedError
+
+
+class ImageClassificationTrainer(Trainer):
+    """
+    Trainer for image classification.
+
+    :param model: The loaded torch model to train.
+    :param key: The arch key of the model.
+    :param recipe_path: The path to the yaml file containing the modifiers and
+        schedule to apply them with; Can also provide a SparseZoo stub prefixed
+        with 'zoo:'.
+    :param local_rank: The local rank of the current process. Should be set
+        by torch and passed to the trainer.
+    :param rank: The ddp rank of the current process. Defaults to -1
+    :param device: The device to train on. Defaults to torch default device.
+    :param use_mixed_precision: Whether to use mixed precision FP16 training.
+        Defaults to False.
+    :param sparse_transfer_learn: True if transfer learning on a sparse model,
+        else False. Defaults to False.
+    :param val_loader: A DataLoader for validation data.
+    :param train_loader: A DataLoader for training data.
+    :param is_main_process: Whether the current process is the main process,
+        while training using DDP. Defaults to True.
+    :param loggers: A list of loggers to use during training process.
+    :param loss_fn: A Callabale loss function for training and validation
+        losses.
+    :param init_lr: The initial learning rate for the optimizer.Defaults to
+        1e-9
+    :param optim_name: str respresenting the optimizer type to use, one of
+        [SGD, Adam, RMSprop]. Defaults to `adam`.
+    :param optim_kwargs: dict of additional kwargs to pass to the optimizer.
+    """
+
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        key: str,
+        recipe_path: str,
+        local_rank: int,
+        rank: int = -1,
+        device: str = default_device(),
+        use_mixed_precision: bool = False,
+        sparse_transfer_learn: bool = False,
+        val_loader: Optional[DataLoader] = None,
+        train_loader: Optional[DataLoader] = None,
+        is_main_process: bool = True,
+        loggers: Optional[List[Any]] = None,
+        loss_fn: Callable = lambda: CrossEntropyLossWrapper(),
+        init_lr=1e-9,
+        optim_name="adam",
+        optim_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        """
+        Initializes the trainer.
+        """
+        # self.checkpoint_path = checkpoint_path
+        self.recipe_path = recipe_path
+        # self.num_classes = num_classes
+        self.local_rank = local_rank
+        self.rank = rank
```

Comment:
let's get rid of rank and local_rank if possible

Reply:
Both `local_rank` and `rank` are needed for moving the model to respective device and `ddp` setup

Reply:
Let's move device selection outside of trainer.

Reply:
Done!

Reply:
rank is still needed for syncing ddp dataloaders during training

Reply:
Removed in favor of passing in a `ddp` flag